### PR TITLE
feature(clients): Preload machine types for GCP

### PR DESCRIPTION
**Pull request by adiabramovitch**
This PR adds the feature of preloading all the machine types in GCP and also refactors the structure of the customer and service client for gcp. 
in a different PR, I will add EC2 refactoring as well. 

It is mostly done, it is in draft mode so I can hear your opinion,maybe some changes are needed. 

Missing(?): 
- Tests 
- OpenAPI
- EC2 refactoring

### CODE REVIEW

**Review by lzap — COMMENTED** [review]
Nice, few comments but small things. I'd appreciate the YAML changes in a separate commit, but here we are, you can keep that.

Now the big question is, can you run some tool that finds duplicate files and use it on the YAML files? Can you find same files? I am wondering if there are same file. Does not look like for other clouds it is the case:

```
% fdupes internal/clients/http/*/types/availability/ 
internal/clients/http/ec2/types/availability/us-west-2.yaml
internal/clients/http/ec2/types/availability/us-east-1.yaml

internal/clients/http/azure/types/availability/germanywestcentral_1.yaml
internal/clients/http/azure/types/availability/germanywestcentral_2.yaml

internal/clients/http/azure/types/availability/norwayeast_1.yaml
internal/clients/http/azure/types/availability/eastasia_3.yaml
internal/clients/http/azure/types/availability/uaenorth_3.yaml

internal/clients/http/azure/types/availability/qatarcentral_3.yaml
internal/clients/http/azure/types/availability/qatarcentral_1.yaml
```

File: internal/clients/interface.go [hunk 1]
```diff
@@ -82,15 +83,28 @@ type Azure interface {
...
+type GCPService interface {
+	// RegisterInstanceTypes
+	RegisterInstanceTypes(ctx context.Context, instanceTypes *RegisteredInstanceTypes, regionalTypes *RegionalTypeAvailability) error
+
+	// ListMachineTypes returns list of all GCP machine types
+	ListMachineTypes(ctx context.Context, zone string) ([]*InstanceType, error)
 
 	// ListAllRegionsAndZones returns list of all GCP regions
 	ListAllRegionsAndZones(ctx context.Context) ([]Region, []Zone, error)
+}
```
[the remark this answers is not in the document]
Comment (adiabramovitch) [reply internal/clients/interface.go:101]: forgot, Added it

File: internal/clients/http/gcp/gcp_service_client.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+	"github.com/RHEnVision/provisioning-backend/internal/models"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+type gcpServiceClient struct {
+	auth    *clients.Authentication
+	options []option.ClientOption
+}
```
[the remark this answers is not in the document]
Comment (adiabramovitch) [reply internal/clients/http/gcp/gcp_service_client.go:23]: You are right, this is why I added the project id to the GCP config, removing the auth field.

File: internal/clients/http/gcp/gcp_service_client.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+func getMachineFamily(machineType string) string {
+	machineFamily := strings.Split(machineType, "-")
+	return machineFamily[0]
+}
+
+func getTotalStorage(disks []*computepb.ScratchDisks) int64 {
+	var sum int64 = 0
+	for _, disk := range disks {
+		sum = sum + int64(*disk.DiskGb)
```
Comment (avitova) [inline internal/clients/http/gcp/gcp_service_client.go:167]:
```suggestion
		sum += int64(*disk.DiskGb)
```
I think this should work as well and looks cleaner.

**Review by avitova — COMMENTED** [review]
LGTM. (Even though I am a fan of one liners and would maybe try to shrink mbToMiB and getMachineFamily.)

**Comment by adiabramovitch** [comment]
> 

Do you want me to add a make target or a script to check dup files?

**Comment by lzap** [comment]
Oh no, no script needed, I was just curious if GCP maybe has all zones with the very same types. This was not the case for neither EC2 nor Azure. Just wanted to know, I can take a look later no big deal.

File: internal/clients/http/ec2/ec2_client.go [hunk 4]
```diff
@@ -31,7 +31,7 @@ type ec2Client struct {
 }
 
 func init() {
-	clients.GetCustomerEC2Client = newAssumedEC2ClientWithRegion
+	clients.GetEC2Client = newAssumedEC2ClientWithRegion
 	clients.GetServiceEC2Client = newEC2ClientWithRegion
 }
```
Comment (lzap) [inline internal/clients/http/ec2/ec2_client.go:36]: A separate commit would be great, but whatever it's there leave it.

File: internal/clients/http/gcp/gcp_service_client.go (hunk 3 above)
Comment (lzap) [reply internal/clients/http/gcp/gcp_service_client.go:159]: Sure let's do that.

File: internal/clients/interface.go [hunk 5]
```diff
@@ -82,15 +83,28 @@ type Azure interface {
 	ClientStatuser
 }
 
-// GetGCPClient returns a GCP facade interface.
+// GetCustomerGCPClient returns a GCP facade interface of the customer.
```
Comment (lzap) [inline internal/clients/interface.go:86]: This should be `GetGCPClient`

**Review by lzap — CHANGES_REQUESTED** [review]
Tiny typo plus Anna's remark, but looks great.

**Comment by adiabramovitch** [comment]
@lzap, @avitova Thank you for your comments. 
I added Anna's recommendation and broke the commit into a few separate ones.

**Review by lzap — APPROVED** [review]

**Pull request merged**